File: js/src/feistel.ts
import { hmac } from "@noble/hashes/hmac";
import { sha256 } from "@noble/hashes/sha256";
import { HrcError } from "./errors.js";

const TAG = new TextEncoder().encode("HRC-FEISTEL-V1");
const MAX_WALKS = 1000;

interface FeistelKey {
  profileId: string;
  keyBytes: Uint8Array;
  rounds: number;
}

function bitLength(capacity: bigint): number {
  return (capacity - 1n).toString(2).length;
}

/** Low n bits of the HMAC-SHA-256 digest, per spec 7.3. */
function lowBits(digest: Uint8Array, n: number): bigint {
  const byteCount = Math.ceil(n / 8);
  let v = 0n;
  for (let i = 0; i < byteCount; i += 1) {
    v = (v << 8n) | BigInt(digest[i] as number);
  }
  return v & ((1n << BigInt(n)) - 1n);
}

function toBe(value: bigint, byteCount: number): Uint8Array {
  const out = new Uint8Array(byteCount);
  let v = value;
  for (let i = byteCount - 1; i >= 0; i -= 1) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

function roundMessage(profileId: string, round: number, right: bigint, wr: number): Uint8Array {
  const pidBytes = new TextEncoder().encode(profileId);
  const rightBytes = toBe(right, Math.ceil(wr / 8));
  const msg = new Uint8Array(TAG.length + 1 + pidBytes.length + 1 + 1 + rightBytes.length);
  let o = 0;
  msg.set(TAG, o); o += TAG.length;
  msg[o] = 0; o += 1;
  msg.set(pidBytes, o); o += pidBytes.length;
  msg[o] = 0; o += 1;
  msg[o] = round; o += 1;
  msg.set(rightBytes, o);
  return msg;
}

interface Halves {
  left: bigint;
  right: bigint;
}

function runRounds(h: Halves, key: FeistelKey, w0: number, w1: number): Halves {
  let { left, right } = h;
  for (let i = 0; i < key.rounds; i += 1) {
    const even = i % 2 === 0;
    const wr = even ? w1 : w0;
    const wl = even ? w0 : w1;
    const digest = hmac(sha256, key.keyBytes, roundMessage(key.profileId, i, right, wr));
    const f = lowBits(digest, wl);
    const newLeft = right;
    const newRight = left ^ f;
    left = newLeft;
    right = newRight;
  }
  return { left, right };
}

function runInverse(h: Halves, key: FeistelKey, w0: number, w1: number): Halves {
  let { left, right } = h;
  for (let i = key.rounds - 1; i >= 0; i -= 1) {
    const even = i % 2 === 0;
    const wr = even ? w1 : w0;
    const wl = even ? w0 : w1;
    const digest = hmac(sha256, key.keyBytes, roundMessage(key.profileId, i, left, wr));
    const f = lowBits(digest, wl);
    const prevRight = left;
    const prevLeft = right ^ f;
    left = prevLeft;
    right = prevRight;
  }
  return { left, right };
}

function combine(h: Halves, w1: number): bigint {
  return (h.left << BigInt(w1)) | h.right;
}

function split(value: bigint, w1: number): Halves {
  return { left: value >> BigInt(w1), right: value & ((1n << BigInt(w1)) - 1n) };
}

/** Spec 7.3 forward permutation with cycle walking. */
export function permute(value: bigint, capacity: bigint, key: FeistelKey): bigint {
  const bits = bitLength(capacity);
  const w1 = Math.floor(bits / 2);
  const w0 = bits - w1;
  let v = value;
  for (let walk = 0; walk < MAX_WALKS; walk += 1) {
    const out = combine(runRounds(split(v, w1), key, w0, w1), w1);
    if (out < capacity) return out;
    v = out;
  }
  throw new HrcError("PERMUTATION_FAILURE", "Feistel cycle walking exceeded 1000 iterations", false);
}

/** Spec 7.3 inverse permutation with cycle walking. */
export function inversePermute(value: bigint, capacity: bigint, key: FeistelKey): bigint {
  const bits = bitLength(capacity);
  const w1 = Math.floor(bits / 2);
  const w0 = bits - w1;
  let v = value;
  for (let walk = 0; walk < MAX_WALKS; walk += 1) {
    const out = combine(runInverse(split(v, w1), key, w0, w1), w1);
    if (out < capacity) return out;
    v = out;
  }
  throw new HrcError("PERMUTATION_FAILURE", "Feistel cycle walking exceeded 1000 iterations", false);
}

File: rust/src/feistel.rs
//! Feistel-v1 reversible permutation (spec section 7.3).
//!
//! Balanced Feistel network with alternating half widths, an HMAC-SHA-256
//! round function truncated to the low N bits and cycle walking to keep
//! outputs inside `0 .. capacity - 1`.

use hmac::{Hmac, Mac};
use num_bigint::BigUint;
use sha2::Sha256;

use crate::error::{BasehError, ErrorCode};

const TAG: &[u8] = b"BASEH-FEISTEL-V1";
const MAX_WALKS: u32 = 1000;

type HmacSha256 = Hmac<Sha256>;

/// Number of bits needed to represent `capacity - 1` (`ceil(log2(capacity))`
/// for capacity >= 2).
fn bit_length(capacity: &BigUint) -> usize {
    let one = BigUint::from(1u64);
    (capacity - one).bits() as usize
}

/// Low n bits of the digest: first ceil(n / 8) bytes read big-endian,
/// masked with `2^n - 1`.
fn low_bits(digest: &[u8], n: usize) -> BigUint {
    let byte_count = n.div_ceil(8);
    let mut v = BigUint::from(0u64);
    for byte in digest.iter().take(byte_count) {
        v = (v << 8usize) | BigUint::from(*byte);
    }
    if n == 0 {
        return BigUint::from(0u64);
    }
    let mask = (BigUint::from(1u64) << n) - 1u64;
    v & mask
}

/// Unsigned big-endian encoding in exactly `byte_count` bytes.
fn to_be(value: &BigUint, byte_count: usize) -> Vec<u8> {
    let raw = value.to_bytes_be();
    if raw.len() >= byte_count {
        raw[raw.len() - byte_count..].to_vec()
    } else {
        let mut out = vec![0u8; byte_count - raw.len()];
        out.extend_from_slice(&raw);
        out
    }
}

fn round_message(profile_id: &str, round: u32, right: &BigUint, wr: usize) -> Vec<u8> {
    let right_bytes = to_be(right, wr.div_ceil(8));
    let mut msg = Vec::with_capacity(TAG.len() + 1 + profile_id.len() + 1 + 1 + right_bytes.len());
    msg.extend_from_slice(TAG);
    msg.push(0);
    msg.extend_from_slice(profile_id.as_bytes());
    msg.push(0);
    msg.push(round as u8);
    msg.extend_from_slice(&right_bytes);
    msg
}

struct Key<'a> {
    profile_id: &'a str,
    key_bytes: &'a [u8],
    rounds: u32,
}

struct Halves {
    left: BigUint,
    right: BigUint,
}

fn round_f(key: &Key<'_>, round: u32, value: &BigUint, wr: usize, wl: usize) -> BigUint {
    let mut mac = <HmacSha256 as Mac>::new_from_slice(key.key_bytes)
        .expect("HMAC accepts keys of any length");
    mac.update(&round_message(key.profile_id, round, value, wr));
    let digest = mac.finalize().into_bytes();
    low_bits(&digest, wl)
}

fn widths(i: u32, w0: usize, w1: usize) -> (usize, usize) {
    if i.is_multiple_of(2) {
        (w1, w0)
    } else {
        (w0, w1)
    }
}

fn run_rounds(h: Halves, key: &Key<'_>, w0: usize, w1: usize) -> Halves {
    let Halves {
        mut left,
        mut right,
    } = h;
    for i in 0..key.rounds {
        let (wr, wl) = widths(i, w0, w1);
        let f = round_f(key, i, &right, wr, wl);
        let new_left = right;
        let new_right = left ^ &f;
        left = new_left;
        right = new_right;
    }
    Halves { left, right }
}

fn run_inverse(h: Halves, key: &Key<'_>, w0: usize, w1: usize) -> Halves {
    let Halves {
        mut left,
        mut right,
    } = h;
    for i in (0..key.rounds).rev() {
        let (wr, wl) = widths(i, w0, w1);
        let f = round_f(key, i, &left, wr, wl);
        let prev_right = left;
        let prev_left = right ^ &f;
        left = prev_left;
        right = prev_right;
    }
    Halves { left, right }
}

fn combine(h: &Halves, w1: usize) -> BigUint {
    (&h.left << w1) | &h.right
}

fn split(value: &BigUint, w1: usize) -> Halves {
    let mask = (BigUint::from(1u64) << w1) - 1u64;
    Halves {
        left: value >> w1,
        right: value & &mask,
    }
}

fn walk(
    value: &BigUint,
    capacity: &BigUint,
    key: &Key<'_>,
    forward: bool,
) -> Result<BigUint, BasehError> {
    let bits = bit_length(capacity);
    let w1 = bits / 2;
    let w0 = bits - w1;
    let mut v = value.clone();
    for _ in 0..MAX_WALKS {
        let h = split(&v, w1);
        let out = combine(
            &if forward {
                run_rounds(h, key, w0, w1)
            } else {
                run_inverse(h, key, w0, w1)
            },
            w1,
        );
        if out < *capacity {
            return Ok(out);
        }
        v = out;
    }
    Err(BasehError::new(
        ErrorCode::PermutationFailure,
        "Feistel cycle walking exceeded 1000 iterations",
        false,
    ))
}

/// Spec 7.3 forward permutation with cycle walking.
pub fn permute(
    value: &BigUint,
    capacity: &BigUint,
    profile_id: &str,
    key_bytes: &[u8],
    rounds: u32,
) -> Result<BigUint, BasehError> {
    walk(
        value,
        capacity,
        &Key {
            profile_id,
            key_bytes,
            rounds,
        },
        true,
    )
}

/// Spec 7.3 inverse permutation with cycle walking.
pub fn inverse_permute(
    value: &BigUint,
    capacity: &BigUint,
    profile_id: &str,
    key_bytes: &[u8],
    rounds: u32,
) -> Result<BigUint, BasehError> {
    walk(
        value,
        capacity,
        &Key {
            profile_id,
            key_bytes,
            rounds,
        },
        false,
    )
}
